src/macros.jl

```diff
@@ -241,20 +256,20 @@ macro testitem(nm, exs...)
     end
     q = QuoteNode(exs[end])
     esc(quote
-        $store_test_item_setup(
-            $TestItem(
-                Ref(0), $nm, $tags, $default_imports, $setup, $retries,
-                $(String(__source__.file)), $(__source__.line),
-                $gettls(:__RE_TEST_PROJECT__, "."),
-                $q,
-                $TestSetup[],
-                Ref{Int}(0),
-                $DefaultTestSet[],
-                Ref{Int}(0),
-                $PerfStats[],
-                $ScheduledForEvaluation()
-            )
+        let ti = $TestItem(
+            Ref(0), $nm, $tags, $default_imports, $setup, $retries,
+            $(String(__source__.file)), $(__source__.line),
+            $gettls(:__RE_TEST_PROJECT__, "."),
+            $q,
         )
+            if $_quote || $gettls(:__RE_TEST_RUNNING__, false)::$Bool
+                $store_test_item_setup(ti)
+                ti
+            else # We are not in a `runtests` call, so we run the testitem eagerly.
+                $runtestitem(ti)
+                nothing
```

Comment:
this is the main change

btw, how do people usually assign to variables in a macro if escaping the whole thing? i've used a `let` block here but idk what best practice is... (should we change to just `esc`ing what's necessary?)

Reply:
I think the main approach is to manually gensym a variable name outside the `esc`aped expression, and then interpolate that in everywhere.. it's a bit ugly. Like you point out, top-level escape can get ugly fast.

A let-block seems like a prudent approach (and probably a good idea anyway to scope the object nicely).. I guess the main concern would be if inside any user-provided expressions we might actually capture `ti` if the user used a variable of the same name.

Like if they have a global variable called `ti = 5` and they try to use it to name their testitem `@testitem "run-$ti"`, then it would end up using your let block instead! 😮 

So I think probably you _should_ gensym a new name for `ti` and use that, or switch to narrow escapes.

Reply:
e.g. `ti = gensym("ti")` and then you use `$ti` everywhere in the quote.